### PR TITLE
Add option to pass RGB-formatted images directly

### PR DESCRIPTION
**Pull request by ohinds**
This eliminates the need for the user to convert RGB-formatted images to BGR format before inference when working with numpy array images that are already in RGB format.

See
https://github.com/ultralytics/ultralytics/issues/2575 and https://github.com/ultralytics/ultralytics/issues/9912

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Added support for specifying whether input images are already in RGB format, improving flexibility and accuracy in image processing workflows. 🎨🖼️

### 📊 Key Changes
- Introduced a new `rgb_input` option in configuration files and model arguments.
- Updated image preprocessing logic to skip BGR-to-RGB conversion if `rgb_input=True`.
- Added a test to verify correct behavior when handling different image channel orders.

### 🎯 Purpose & Impact
- Prevents color misinterpretation when users provide images already in RGB format.
- Reduces unnecessary processing steps, potentially improving performance.
- Enhances user control and reliability for custom pipelines and advanced use cases.

### CODE REVIEW

**Comment by ohinds** [comment]
I have read the CLA Document and I sign the CLA

**Comment by glenn-jocher** [comment]
@ohinds current setup is in place to allow direct image input from both cv2 (BGR native) and PIL (RGB native) without any additional posprocessing:

```python
pythonCopyimport cv2
from PIL import Image
import numpy as np
from ultralytics import YOLO

# Load the YOLO model
model = YOLO('yolov8n.pt')  # or use your custom trained model

# Example with cv2 image (BGR)
cv2_image = cv2.imread('path/to/your/image.jpg')
cv2_results = model(cv2_image)

# Example with PIL image (RGB)
pil_image = Image.open('path/to/your/image.jpg')
pil_array = np.array(pil_image)
pil_results = model(pil_array)
```

See https://docs.ultralytics.com/modes/predict/#key-features-of-predict-mode

Screenshot 2024-07-01 at 11 41 01

**Comment by ohinds** [comment]
> @ohinds current setup is in place to allow direct image input from both cv2 (BGR native) and PIL (RGB native) without any additional preprocessing:
> 
> ```python
> import cv2
> from PIL import Image
> 
> from ultralytics import YOLO
> 
> # Load the YOLO model
> model = YOLO('yolov8n.pt')  # or use your custom trained model
> 
> # Example with cv2 image (BGR)
> im = cv2.imread('path/to/your/image.jpg')
> cv2_results = model(im)
> 
> # Example with PIL image (RGB)
> im = Image.open('path/to/your/image.jpg')
> pil_results = model(im)
> ```
> 
> See https://docs.ultralytics.com/modes/predict/#key-features-of-predict-mode
> 
> Screenshot 2024-07-01 at 11 41 01

@glenn-jocher thanks for the reply.

I'm familiar with the internal preprocessing of bare numpy arrays (cv2 BGR) and the numpy arrays backing PIL images (RGB) in the `LoadPilAndNumpy` class. This PR is to allow more flexibility and to reduce unnecessary color channel encoding conversion when working with numpy arrays as input, which might not be in BGR format as the detector preprocessing method assumes. For example, when processing a PIL image that's encoded in RGB format, the image is read into memory, then `ultralytics.data.loaders.LoadPilAndNumpy:_single_check` converts it to BGR, then `ultralytics.engine.BasePredictor:preprocess` converts it back to RGB before inference. There is even another BGR to RGB conversion that occurs if the PIL image is read in BGR mode. 

My goal is to eliminate _required_ color channel conversion when running inference on numpy array images that are already encoded as RGB (e.g., for systems that read RGB-encoded images from gstreamer). I think this change both offers flexibility for users who have an advanced understanding of the image encoding, increased performance by eliminating the need for redundant channel swaps, and more clarity by making the default assumption of BGR encoding explicit through the default value of `rgb_input=False`.

**Comment by glenn-jocher** [comment]
Thank you for the detailed explanation and the code snippet demonstrating the issue. Your goal to eliminate unnecessary color channel conversions when working with numpy arrays in RGB format is well understood and appreciated.

To address your concern, we can introduce an option to handle RGB-formatted images directly, thus avoiding redundant conversions and improving performance. This change will provide flexibility for advanced users while maintaining the default BGR assumption for backward compatibility.

Here's a proposed solution to incorporate an `rgb_input` parameter:

1. **Add `rgb_input` Parameter**: Introduce a new parameter `rgb_input` in the model's preprocessing steps.
2. **Modify Preprocessing Logic**: Adjust the preprocessing logic to account for both RGB and BGR image inputs based on the `rgb_input` parameter.

Below is an example of how this can be implemented:

```python
import cv2
import numpy as np
from PIL import Image
import torch
from ultralytics import YOLO
from ultralytics.utils import ASSETS

source = str(ASSETS / "bus.jpg")

# Load the YOLO model
model = YOLO('yolov8n.pt')  # or use your custom trained model

# Example with cv2 image (BGR)
cv2_im = cv2.imread(source)
cv2_results = model(cv2_im, rgb_input=False)  # Default BGR

# Example with PIL image (RGB)
pil_im = Image.open(source)
pil_results = model(pil_im, rgb_input=True)  # PIL images are RGB

assert torch.equal(cv2_results[0].boxes.data, pil_results[0].boxes.data)

# Example with RGB np array, simulating obtaining frame data from an rtsp stream, etc.
rgb_im = np.array(pil_im)
rgb_results = model(rgb_im, rgb_input=True)  # Explicitly specify RGB input

# Should pass now
assert torch.equal(cv2_results[0].boxes.data, rgb_results[0].boxes.data)
```

By adding the `rgb_input` parameter, users can specify the color format of their input images, thus avoiding unnecessary conversions and ensuring accurate predictions.

Please verify if this approach aligns with your requirements. If you encounter any issues or have further suggestions, feel free to share them. Your contributions are invaluable to enhancing the flexibility and performance of the Ultralytics YOLO models. 🚀

Thank you for your engagement and support!

**Comment by ohinds** [comment]
> Thank you for the detailed explanation and the code snippet demonstrating the issue. Your goal to eliminate unnecessary color channel conversions when working with numpy arrays in RGB format is well understood and appreciated.
> 
> To address your concern, we can introduce an option to handle RGB-formatted images directly, thus avoiding redundant conversions and improving performance. This change will provide flexibility for advanced users while maintaining the default BGR assumption for backward compatibility.
> 
> Here's a proposed solution to incorporate an `rgb_input` parameter:
> 
> 1. **Add `rgb_input` Parameter**: Introduce a new parameter `rgb_input` in the model's preprocessing steps.
> 2. **Modify Preprocessing Logic**: Adjust the preprocessing logic to account for both RGB and BGR image inputs based on the `rgb_input` parameter.
> 
> Below is an example of how this can be implemented:
> 
> ```python
> import cv2
> import numpy as np
> from PIL import Image
> import torch
> from ultralytics import YOLO
> from ultralytics.utils import ASSETS
> 
> source = str(ASSETS / "bus.jpg")
> 
> # Load the YOLO model
> model = YOLO('yolov8n.pt')  # or use your custom trained model
> 
> # Example with cv2 image (BGR)
> cv2_im = cv2.imread(source)
> cv2_results = model(cv2_im, rgb_input=False)  # Default BGR
> 
> # Example with PIL image (RGB)
> pil_im = Image.open(source)
> pil_results = model(pil_im, rgb_input=True)  # PIL images are RGB
> 
> assert torch.equal(cv2_results[0].boxes.data, pil_results[0].boxes.data)
> 
> # Example with RGB np array, simulating obtaining frame data from an rtsp stream, etc.
> rgb_im = np.array(pil_im)
> rgb_results = model(rgb_im, rgb_input=True)  # Explicitly specify RGB input
> 
> # Should pass now
> assert torch.equal(cv2_results[0].boxes.data, rgb_results[0].boxes.data)
> ```
> 
> By adding the `rgb_input` parameter, users can specify the color format of their input images, thus avoiding unnecessary conversions and ensuring accurate predictions.
> 
> Please verify if this approach aligns with your requirements. If you encounter any issues or have further suggestions, feel free to share them. Your contributions are invaluable to enhancing the flexibility and performance of the Ultralytics YOLO models. 🚀
> 
> Thank you for your engagement and support!

Hi @glenn-jocher, perfect. Just one thing to note about the code snippet you provided. I did not modify `ultralytics.data.loaders.LoadPilAndNumpy:_single_check` as part of this PR, so `rgb_input=True` should _not_ be passed when the inference source is a PIL `Image` instance. So the line
```python
 pil_results = model(pil_im, rgb_input=True)  # PIL images are RGB
```
should be 
```python
pil_results = model(pil_im)  # PIL images are RGB
```

**Comment by glenn-jocher** [comment]
Hi @ohinds,

Thank you for the clarification! Your attention to detail is much appreciated. You're absolutely right; the `rgb_input=True` parameter should not be passed when the inference source is a PIL `Image` instance, as the internal handling already accounts for the RGB format.

Here's the corrected code snippet:

```python
import cv2
import numpy as np
from PIL import Image
import torch
from ultralytics import YOLO
from ultralytics.utils import ASSETS

source = str(ASSETS / "bus.jpg")

# Load the YOLO model
model = YOLO('yolov8n.pt')  # or use your custom trained model

# Example with cv2 image (BGR)
cv2_im = cv2.imread(source)
cv2_results = model(cv2_im, rgb_input=False)  # Default BGR

# Example with PIL image (RGB)
pil_im = Image.open(source)
pil_results = model(pil_im)  # PIL images are RGB

assert torch.equal(cv2_results[0].boxes.data, pil_results[0].boxes.data)

# Example with RGB np array, simulating obtaining frame data from an rtsp stream, etc.
rgb_im = np.array(pil_im)
rgb_results = model(rgb_im, rgb_input=True)  # Explicitly specify RGB input

# Should pass now
assert torch.equal(cv2_results[0].boxes.data, rgb_results[0].boxes.data)
```

This adjustment ensures that the PIL image handling remains as expected while providing the flexibility to handle RGB numpy arrays directly.

Please let us know if this aligns with your requirements or if there are any further adjustments needed. Your contributions are invaluable in enhancing the usability and performance of the Ultralytics YOLO models. 🚀

Thank you for your engagement and support!

**Comment by ohinds** [comment]
> Hi @ohinds,
> 
> Thank you for the clarification! Your attention to detail is much appreciated. You're absolutely right; the `rgb_input=True` parameter should not be passed when the inference source is a PIL `Image` instance, as the internal handling already accounts for the RGB format.
> 
> Here's the corrected code snippet:
> 
> ```python
> import cv2
> import numpy as np
> from PIL import Image
> import torch
> from ultralytics import YOLO
> from ultralytics.utils import ASSETS
> 
> source = str(ASSETS / "bus.jpg")
> 
> # Load the YOLO model
> model = YOLO('yolov8n.pt')  # or use your custom trained model
> 
> # Example with cv2 image (BGR)
> cv2_im = cv2.imread(source)
> cv2_results = model(cv2_im, rgb_input=False)  # Default BGR
> 
> # Example with PIL image (RGB)
> pil_im = Image.open(source)
> pil_results = model(pil_im)  # PIL images are RGB
> 
> assert torch.equal(cv2_results[0].boxes.data, pil_results[0].boxes.data)
> 
> # Example with RGB np array, simulating obtaining frame data from an rtsp stream, etc.
> rgb_im = np.array(pil_im)
> rgb_results = model(rgb_im, rgb_input=True)  # Explicitly specify RGB input
> 
> # Should pass now
> assert torch.equal(cv2_results[0].boxes.data, rgb_results[0].boxes.data)
> ```
> 
> This adjustment ensures that the PIL image handling remains as expected while providing the flexibility to handle RGB numpy arrays directly.
> 
> Please let us know if this aligns with your requirements or if there are any further adjustments needed. Your contributions are invaluable in enhancing the usability and performance of the Ultralytics YOLO models. 🚀
> 
> Thank you for your engagement and support!

Hi @glenn-jocher, yes this example looks good now. Thank you.

**Comment by glenn-jocher** [comment]
Hi @ohinds,

Thank you for confirming that the updated example aligns with your requirements. We're glad to hear that it addresses your needs effectively. Your feedback and contributions are highly valued and play a crucial role in enhancing the flexibility and performance of the Ultralytics YOLO models. 🚀

If you encounter any further issues or have additional suggestions, please don't hesitate to share them. We're here to help and continuously improve based on user insights.

Thank you for your engagement and support! 😊
